File: forge/src/main/java/org/waste/of/time/WorldToolsForge.java
package org.waste.of.time;

import net.minecraft.block.entity.BlockEntityType;
import net.minecraft.client.render.block.entity.BlockEntityRendererFactories;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.client.event.ClientPlayerNetworkEvent;
import net.minecraftforge.client.event.RegisterClientCommandsEvent;
import net.minecraftforge.client.event.RegisterKeyMappingsEvent;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.entity.player.PlayerInteractEvent;
import net.minecraftforge.eventbus.api.IEventBus;
import net.minecraftforge.fml.DistExecutor;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import org.waste.of.time.command.WorldToolsForgeCommandBuilder;
import org.waste.of.time.event.Events;
import org.waste.of.time.renderer.MissingChestBlockEntityRenderer;

@Mod("worldtools")
public class WorldToolsForge {
    public static IEventBus FORGE_EVENT_BUS = MinecraftForge.EVENT_BUS;

    public WorldToolsForge() {
        WorldTools.INSTANCE.initialize();
        IEventBus modEventBus = FMLJavaModLoadingContext.get().getModEventBus();
        DistExecutor.safeRunWhenOn(Dist.CLIENT, () -> () -> {
            modEventBus.addListener(this::onInitializeClient);
            FORGE_EVENT_BUS.addListener(this::onRegisterCommands);
            FORGE_EVENT_BUS.addListener(this::onRegisterKeybinds);
            FORGE_EVENT_BUS.addListener(this::onPlayerJoin);
            FORGE_EVENT_BUS.addListener(this::onRightClickBlock);
            BlockEntityRendererFactories.register(BlockEntityType.CHEST, MissingChestBlockEntityRenderer::new);
            FORGE_EVENT_BUS.register(modEventBus);
        });

        WorldTools.INSTANCE.getLOGGER().info("WorldTools Forge initialized");
    }

    public void onInitializeClient(final FMLClientSetupEvent event) {
        WorldTools.INSTANCE.initialize();
    }

    public void onRegisterCommands(final RegisterClientCommandsEvent event) {
        WorldToolsForgeCommandBuilder.INSTANCE.register(event.getDispatcher());
    }

    public void onRegisterKeybinds(final RegisterKeyMappingsEvent event) {
        event.register(WorldTools.INSTANCE.getGUI_KEY());
    }

    public void onPlayerJoin(final ClientPlayerNetworkEvent.LoggingIn event) {
        Events.INSTANCE.onClientJoin();
    }

    public void onRightClickBlock(final PlayerInteractEvent.RightClickBlock event) {
        Events.INSTANCE.onInteractBlock(event.getEntity(), event.getLevel(), event.getHand(), event.getHitVec());
    }
}

File: forge/src/main/kotlin/org/waste/of/time/command/WorldToolsForge.kt
package org.waste.of.time.command

import net.minecraft.block.entity.BlockEntityType
import net.minecraft.client.render.block.entity.BlockEntityRendererFactories
import net.minecraftforge.client.event.ClientPlayerNetworkEvent
import net.minecraftforge.client.event.RegisterClientCommandsEvent
import net.minecraftforge.client.event.RegisterKeyMappingsEvent
import net.minecraftforge.client.event.ScreenEvent
import net.minecraftforge.event.entity.EntityJoinLevelEvent
import net.minecraftforge.event.entity.player.PlayerInteractEvent
import net.minecraftforge.fml.common.Mod
import org.waste.of.time.WorldTools
import org.waste.of.time.event.Events
import org.waste.of.time.renderer.MissingChestBlockEntityRenderer
import thedarkcolour.kotlinforforge.forge.FORGE_BUS

@Mod(WorldTools.MOD_ID)
object WorldToolsForge {
    init {
        WorldTools.initialize()
        FORGE_BUS.addListener<RegisterClientCommandsEvent> {
            WorldToolsForgeCommandBuilder.register(it.dispatcher)
        }
        FORGE_BUS.addListener<RegisterKeyMappingsEvent> {
            it.register(WorldTools.GUI_KEY)
        }
        FORGE_BUS.addListener<ClientPlayerNetworkEvent.LoggingIn> {
            Events.onClientJoin()
        }
        FORGE_BUS.addListener<PlayerInteractEvent.RightClickBlock> {
            Events.onInteractBlock(it.entity, it.level, it.hand, it.hitVec)
        }
        FORGE_BUS.addListener<EntityJoinLevelEvent> {
            Events.onEntityLoad(it.entity)
        }
        FORGE_BUS.addListener<ScreenEvent.Init.Post> {
            Events.onAfterInitScreen(it.screen)
        }
        BlockEntityRendererFactories.register(BlockEntityType.CHEST) {
            MissingChestBlockEntityRenderer(it)
        }
    }
}